### PR TITLE
drivers: i3c: cdns i3c driver

### PR DESCRIPTION
**Pull request by XenuIsWatching**
This is initial support of the i3c cdns controller.
This also fixes few issues that were found during development. Please view their commit message for details.

### CODE REVIEW

File: drivers/i3c/Kconfig.cdns [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+DT_COMPAT_CADENCE_I3C := cdns,i3c
+
+module = I3C_CADENCE
+module-str = i3c-cadence
+source "subsys/logging/Kconfig.template.log_config"
+
+config I3C_CADENCE
+	bool "Cadence I3C driver"
+	select I3C_IBI_WORKQUEUE if I3C_USE_IBI
+	default y
```
[the remark this answers is not in the document]
Comment (XenuIsWatching) [reply drivers/i3c/Kconfig.cdns:13]: fixed

File: drivers/i3c/Kconfig.cdns [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2022 Meta Platforms, Inc. and its affiliates.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_CADENCE_I3C := cdns,i3c
```
[the remark this answers is not in the document]
Comment (XenuIsWatching) [reply drivers/i3c/Kconfig.cdns]: fixed

File: drivers/i3c/i3c_cdns.c [hunk 3]
```diff
@@ -0,0 +1,2459 @@
...
+static int cdns_i3c_read_rx_fifo(const struct cdns_i3c_config *config, void *buf, uint32_t len)
+{
+	uint32_t *ptr = buf;
+	uint32_t remain, val;
+
+	for (remain = len; remain >= 4; remain -= 4) {
+		if (cdns_i3c_rx_fifo_empty(config)) {
+			return -EIO;
+		}
+		val = sys_cpu_to_le32(sys_read32(config->base + RX_FIFO));
```
[the remark this answers is not in the document]
Comment (XenuIsWatching) [reply drivers/i3c/i3c_cdns.c]: fixed

File: drivers/i3c/i3c_cdns.c [hunk 4]
```diff
@@ -0,0 +1,2459 @@
...
+	struct cdns_i3c_data *data = dev->data;
+
+	/* Clear all retaining regs */
+	sys_write32(DEVS_CTRL_DEV_CLR_ALL, config->base + DEVS_CTRL);
+
+	uint32_t dev_id_rr0;
+	uint32_t dev_id_rr1;
+	uint32_t dev_id_rr2;
+
+	/* program I3C devices */
```
[the remark this answers is not in the document]
Comment (XenuIsWatching) [reply drivers/i3c/i3c_cdns.c:759]: fixed

File: drivers/i3c/i3c_common.c [hunk 5]
```diff
@@ -402,23 +402,13 @@ static int i3c_bus_setdasa(const struct device *dev,
...
-		}
-
 		LOG_DBG("SETDASA for 0x%x", desc->static_addr);
 
 		ret = i3c_ccc_do_setdasa(desc);
 		if (ret == 0) {
-			desc->dynamic_addr = desc->static_addr;
+			desc->dynamic_addr = (desc->init_dynamic_addr ? desc->init_dynamic_addr
+								      : desc->static_addr);
+			i3c_reattach_i3c_device(dev, desc, desc->static_addr);
```
Comment (dcpleung) [inline drivers/i3c/i3c_common.c:411]: Need to handle when this call returns errors.

Comment (XenuIsWatching) [reply drivers/i3c/i3c_common.c:411]: This was something that I did think about.... but I decided not to do anything if there is an error here as this api is optional and could return ENOSYS..... but there is the issue of if that address is not available.

I do think this reattach api could be made to handle the addr_list before calling in to the driver implemented function.... but that would require some refactoring of how i3c controllers have their config/data commomly defined. I wrote a bit about it in this issue https://github.com/zephyrproject-rtos/zephyr/issues/51828

**Review by dcpleung — DISMISSED** [review]

**Comment by XenuIsWatching** [comment]
I made a small change to the i3c_reattach_i3c_device removing a redundant parameter.

**Review by dcpleung — DISMISSED** [review]

**Comment by XenuIsWatching** [comment]
I found a single header that was 'auto-included' by my IDE that shouldn't been there in `i3c_cdns.c`. I've removed it and did a rebase.

**Review by cfriedt — DISMISSED** [review]

**Comment by dcpleung** [comment]
Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).

**Comment by XenuIsWatching** [comment]
> Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).

No, there aren't any that I know of :( this is all intended with custom ASICs and FPGAs for those to use. While I understand your point... I'm not sure that it is general policy, but it probably should be. There ought to be something like `/tests/drivers/build_all/...` for these cases

**Comment by XenuIsWatching** [comment]
> I found a single header that was 'auto-included' by my IDE that shouldn't been there in `i3c_cdns.c`. I've removed it and did a rebase.

Apparently, I was wrong. That header was added by "me", and was to be there

**Comment by cfriedt** [comment]
> Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).

@teburd, @dcpleung - we should probably have some entries under `tests/drivers/build_all` for i2c / i3c (build-only obviously). Likely out of scope for this PR.

**Review by cfriedt — DISMISSED** [review]

**Comment by dcpleung** [comment]
> > Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).
> 
> @teburd, @dcpleung - we should probably have some entries under `tests/drivers/build_all` for i2c / i3c (build-only obviously). Likely out of scope for this PR.

I think it would be a good idea to do that for this driver. The NXP I3C driver is being built via sensor sample so that is covered.

**Comment by XenuIsWatching** [comment]
> @XenuIsWatching - for reference, I just made an fpga build all test under `tests/drivers/build_all/fpga`. Feel free to ping if you need help.

Thanks, I think I figured it out. I just created a new folder, but it builds under the `qemu_cortex_m3` board as that seems generic enough.

**Review by cfriedt — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**